### ISSUE TITLE
Crash on profile registration when compiled with libstdc++ 4.8

### ISSUE THREAD

**Issue by mbunkin** [LubosD/twinkle#31]
After authorising successfully with the SIP server, twinkle crashes. This is the complete output in the console:
Language name: "en" 
terminate called after throwing an instance of 'std::regex_error'
  what():  regex_error
Aborted (core dumped)

Another case was reported by the user which I did not witness, so I cannot tell when it happened exactly:
Language name: "en" 
Qt has caught an exception thrown from an event handler. Throwing
exceptions from an event handler is not supported in Qt. You must
reimplement QApplication::notify() and catch all exceptions there.

Segmentation fault (core dumped)

The system on which it is run:
Linux 112 3.16.0-50-generic #67~14.04.1-Ubuntu SMP Fri Oct 2 22:07:51
UTC 2015 x86_64 x86_64 x86_64 GNU/Linux

Elementary OS Freya (GNU/Linux 3.16.0-43-generic x86_64)

Last message in the Twinkle log:
+++ 27-10-2015 18:02:58.723583 INFO SIP ::send_sip_udp
Send to: udp:192.168.1.6:5060
PUBLISH sip:XXX@192.168.1.6 SIP/2.0
Via: SIP/2.0/UDP 192.168.1.109;rport;branch=z9hG4bKeisndyxh
Max-Forwards: 70
To: "XXX" sip:XXX@192.168.1.6
From: "XXX" sip:XXX@192.168.1.6;tag=oyycx
Call-ID: ugumicprfumiyjy@XXX
CSeq: 65 PUBLISH
Content-Type: application/pidf+xml
Event: presence
Expires: 3600
User-Agent: Twinkle/1.9.0
Content-Length: 189

<?xml version="1.0" encoding="UTF-8"?>
<presence xmlns="urn:ietf:params:xml:ns:pidf" entity="sip:XXX@192.168.1.6"><tuple id="ekptdp"><status><basic>open</basic></status></tuple></presence>

Cannot find the dumped core to analyse it. I guess it tries to run regex on received response and fails? But twinkle-console does not crash, registers alright.
Will appreciate any suggestion. Thank you!

**Comment by LubosD** [comment]
I'm afraid you'll need to attach gdb... something like this:

````
$ gdb twinkle
(gdb) catch throw
(gdb) r
... do what causes the exception ...
(gdb) bt
````

**Comment by mbunkin** [comment]
Hello,

I’ll do that, but I found the place where it breaks and put it into try {} catch - it’s regular expressions in src/mwi/simple_msg_sum_body.cpp
It was the only place with regex without try {} catch wrap.
I’ll try to figure out what gives the problem, probably print out the regular expression on catch, etc, and let you know.

Best regards,
Michael.

**Comment by mbunkin** [comment]
On crash, the s variable in t_msg_summary::parse(const string &s) contained this value:
0/0 (0/0)

It broke regexp, apparently. Let me know if it helps or if you need more info.

**Comment by LubosD** [comment]
Hmm, std::regex_error should only be thrown if the regexp itself is broken, not the input string (s). Which version of GCC are you using? std::regex is working only since GCC 4.9.0.

**Comment by mbunkin** [comment]
I see. My GCC is 4.8.6, it’s pretty tricky to install GCC 4.9 on Ubuntu, even though I have 14.04 - is there any workaround you could suggest?

**Comment by LubosD** [comment]
Not really, since the regex implementation in GCC <4.9 is completely missing. It just throws exceptions :-(

You could maybe try installing libc++ (alternative to libstdc++) and then add "-stdlib=libc++" into CXXFLAGS for Twinkle build (into CMakeLists.txt next to -std=c++11).

**Comment by mbunkin** [comment]
Thank you very much, I will try and report the results.

**Comment by mkubecek** [comment]
If you do not want to touch your toolchain, you can also try the two revert patches I used for openSUSE packages (versions with gcc 4.8). See (https://build.opensuse.org/package/show/openSUSE:Factory/twinkle), it's the two with names starting with "Revert". But you will need boost then.

**Comment by fbriere** [comment]
For the record, linking against libc++ is not really a practical solution, unless you're willing to recompile all of Qt as well; `std::` objects of libc++ and libstdc++ have a different internal representation and are not interchangeable.

(For the curious, there's a [nice explanation on StackOverflow](http://stackoverflow.com/a/12546314) about how linker errors are used to prevent such substitutions.  But AFAIK, it only works for function arguments, and not return values; this probably explains why a libc++ Twinkle *will* link against a libstdc++ Qt, even though it will most likely crash at runtime.)

@mbunkin: In case you are still running on _trusty_, do know that there's a [PPA available with gcc 4.9](https://launchpad.net/~ubuntu-toolchain-r/+archive/ubuntu/test).
